### PR TITLE
Feature/dhfprod 1885 1889

### PR DESCRIPTION
**Pull request by srinathgit**
This PR

1. Handles exceptions that happen during step run and writes the step output events to job doc
2. Uses new options object when starting each step and when adding uris to options in the step
3. Writes the status of steps that completed with errors as "completed with error step <stepnum>"
4. Return FlowStepModel to UI rather than Flow
5. Minor null checks

### CODE REVIEW

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java [hunk 1]
```diff
@@ -255,10 +255,10 @@ public Flow runFlow(String flowName, List<String> steps) {
             steps.forEach((step) -> restrictedSteps.add(this.getStepKeyInStepMap(flow, step)));
             resp = flowRunner.runFlow(flowName, restrictedSteps);
         }
-        return flowManager.getFlow(flowName);
+        return FlowStepModel.transformFromFlow(flowManager.getFlow(flowName));
```
Comment (hao1st) [inline web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: you can call getFlow(flowName) directly without using flowManager

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java [hunk 2]
```diff
@@ -267,7 +267,7 @@ public Flow stop(String flowName) {
         while(itr.hasNext()){
             flowRunner.stopJob(itr.next());
         }
-        return flowManager.getFlow(flowName);
+        return FlowStepModel.transformFromFlow(flowManager.getFlow(flowName));
```
Comment (hao1st) [inline web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: you can call getFlow(flowName) directly without using flowManager

File: marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java [hunk 3]
```diff
@@ -103,15 +104,15 @@ public RunFlowResponse runFlow(String flowName, List<String> stepNums, String jo
             flow.setOverrideDestDB(destDB);
         }
 
-        if(sourceDB == null){
```
Comment (hao1st) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java]: Please use if (StringUtils.isNotEmpty(sourceDB)) {..} the same for other checks.
https://github.com/marklogic/marklogic-data-hub/blob/9fdff17adb1ec88e052ce5de153691fdc70bf01c/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L119

        if (options != null && !options.isEmpty()) {
            flow.setOverrideOptions(options);
        }

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/QueryStepRunner.java [hunk 4]
```diff
@@ -197,10 +197,16 @@ public Job run() {
...
+            job.withSuccess(false);
             StringWriter errors = new StringWriter();
             e.printStackTrace(new PrintWriter(errors));
             job.withJobOutput(errors.toString());
-            jobUpdate.postJobs(jobId, JobStatus.FAILED.toString(), step);
+            try {
+                jobUpdate.postJobs(jobId, JobStatus.FAILED.toString(), step);
+            }
+            catch (Exception ex) {
+                throw (ex);
```
Comment (hao1st) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/QueryStepRunner.java]: throw ex

File: marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java (hunk 3 above)
Comment (srinathgit) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java]: 1. options is a map and can be empty, so no need for the check
2. will use StringUtils.isNotEmpty for sourceDB

Comment (hao1st) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java]: if the map is empty, do you still want to ovveride the Options in Flow?

Comment (srinathgit) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java]: It is not going to affect anything as they will be passed to stepRunner,  which can add its options( if present)  and will be sent to the server where they are combined.

**Pull request merged**